File: sparkcql-core/src/main/scala/gov/va/sparkcql/io/Log.scala
package gov.va.sparkcql.io

import org.slf4j.LoggerFactory
import org.apache.log4j.Logger
import org.apache.log4j.Level
import org.apache.logging.slf4j.Log4jLogger
import org.apache.spark.sql.{DataFrame, Dataset}
import java.io.ByteArrayOutputStream
import org.apache.logging.log4j.{Level, LogManager}

class Log {}

object Log {
  var forceToConsole: Boolean = true
    
  val logger = LoggerFactory.getLogger(classOf[Log])

  def error(msg: String): Unit = {
    logger.error(msg)
    if (forceToConsole) {
      System.err.println(ConsoleColors.BRIGHT_RED + "ERROR: " + msg + ConsoleColors.RESET)
    }
  }

  def warn(msg: String): Unit = {
    logger.warn(msg)
    if (forceToConsole) {
      System.err.println(ConsoleColors.BRIGHT_YELLOW + "WARNING: " + msg + ConsoleColors.RESET)
    }
  }

  def info(msg: String): Unit = {
    logger.info(msg)
    if (forceToConsole) {
      System.err.println(ConsoleColors.BRIGHT_WHITE + msg + ConsoleColors.RESET)
    }
  }

  def info[T](df: Dataset[_]): Unit = {
    info(capture(df))
  }

  def debug(msg: String): Unit = {
    logger.info(msg)
    if (forceToConsole) {
      System.err.println(ConsoleColors.BRIGHT_WHITE + msg + ConsoleColors.RESET)
    }
  }

  def debug[T](df: DataFrame): Unit = {
    debug(capture(df))
  }

  protected def capture[T](df: Dataset[_]): String = {
    if (df != null) {
      val outCapture = new ByteArrayOutputStream
      Console.withOut(outCapture) {
        df.show()
      }
      new String(outCapture.toByteArray)
    } else {
      "DataFrame is empty"
    }
  }
}

File: sparkcql-core/src/main/java/gov/va/sparkcql/common/Log.java
package gov.va.sparkcql.common;

import org.apache.spark.sql.Dataset;
import org.slf4j.LoggerFactory;
import org.slf4j.Logger;
import java.io.ByteArrayOutputStream;
import java.io.PrintStream;

public class Log {
  
    public static Boolean forceToConsole = true;
    public static Logger logger = LoggerFactory.getLogger(Log.class);

    public static void error(String msg) {
        logger.error(msg);
        if (forceToConsole) {
            System.err.println(ConsoleColors.BRIGHT_RED + "ERROR: " + msg + ConsoleColors.RESET);
        }
    }

    public static void warn(String msg) {
        logger.warn(msg);
        if (forceToConsole) {
            System.err.println(ConsoleColors.BRIGHT_YELLOW + "WARNING: " + msg + ConsoleColors.RESET);
        }
    }

    public static void info(String msg) {
        logger.info(msg);
        if (forceToConsole) {
            System.err.println(ConsoleColors.BRIGHT_WHITE + msg + ConsoleColors.RESET);
        }
    }

    public static void info(Dataset<?> df) {
        info(capture(df));
    }

    public static void debug(String msg) {
        logger.info(msg);
        if (forceToConsole) {
            System.err.println(ConsoleColors.BRIGHT_WHITE + msg + ConsoleColors.RESET);
        }
    }

    public static void debug(Dataset<?> ds) {
        debug(capture(ds));
    }

    protected static String capture(Dataset<?> ds) {
        if (ds != null) {
            PrintStream oldStream = System.out;
            try {
                var outStream = new ByteArrayOutputStream();
                PrintStream printStream = new PrintStream(outStream);
                System.setOut(printStream);
                ds.show();
                System.out.flush();
                return outStream.toString();
            } finally {
                System.setOut(oldStream);
            }
        } else {
            return "DataFrame is empty";
        }
    }
}